### PR TITLE
Int 4144

### PR DESCRIPTION
**Pull request by akryvtsun**
- minor formatting and refactoring
- NPE checks for `expression` and `expressionContext`
- unit tests

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@englishman Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-integration&pullRequestId=1940)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-integration&pullRequestId=1940) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/faq) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@englishman Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-integration&pullRequestId=1940)!

File: spring-integration-core/src/main/java/org/springframework/integration/handler/LoggingHandler.java
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/handler/LoggingHandler.java]: OK, @englishman , everything looks good. 
Just several concerns, which might not be critical, but would be better to follow.
Just because that is project style.

You can't call non-`final` methods from ctors: http://stackoverflow.com/questions/7477553/in-java-is-there-a-legitimate-reason-to-call-a-non-final-method-from-a-class-co
And quoting:

> In general it's a bad idea to call a non-final method within a constructor for precisely this reason - the subclass constructor body won't have been executed yet, so you're effectively calling a method in an environment which hasn't been fully initialized.

Since we don't know how wrong `LoggingHandler` is used in the target apps, we can make `setLevel()` `final`, especially for the maintain version `4.3.x`.
Although I guess we can make a `private void doSetLevel()` to overcome that issue :smile:

File: spring-integration-core/src/main/java/org/springframework/integration/handler/LoggingHandler.java
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/handler/LoggingHandler.java]: This is fine during initialization, but sounds strange at runtime.
The `notNull()` throws `IllegalArgumentException` which doesn't reflect reality for the current `message`. 
We should use here `Assert.state()`, because it is really `IllegalStateException` during handling `message`.

OTOH I have changed my mind and decided to have this code after one more review:
```
private volatile Expression expression = EXPRESSION_PARSER.parseExpression("payload");
...
private volatile EvaluationContext evaluationContext = ExpressionUtils.createStandardEvaluationContext();
```
This isn't Spring Context aware, but lets do not have these asserts at all. Plus such a change reinstates the previous behavior. Therefore the next `4.3.5` will let you to upgrade without issue here :smile:

File: spring-integration-core/src/test/java/org/springframework/integration/handler/LoggingHandlerTests.java
Comment (artembilan) [inline spring-integration-core/src/test/java/org/springframework/integration/handler/LoggingHandlerTests.java]: If you follow with my change to the `LoggingHandler`, these tests won't path anymore.
They would need to be changed to a new logic. Although that will cover the logic you have came to us originally.

OTOH we chase a bit different pattern to verify exceptions.
See, both your tests expect the same outcome. I can change a lot of in code and in tests, too, but they will still pass.

To prevent that we do this:
```
try {
	loggingHandler.handleMessage(message);
	fail("Expected MessageHandlingException");
}
catch (Exception e) {
    assertThat(e, instanceOf(MessageHandlingException.class));
    assertThat(e.getCause(), instanceOf(IllegalStateException.class));
    assertThat(e.getMessage(), containsString("'expression' must not be null"));
}
```

**Review by artembilan — CHANGES_REQUESTED** [review]
And a general comment:
To make a change properly managed and traceable, it would be better to say in commit message the reason of the change and have a link for JIRA and GH issue.

Thank you for the contribution!

File: spring-integration-core/src/main/java/org/springframework/integration/handler/LoggingHandler.java
Comment (englishman) [inline spring-integration-core/src/main/java/org/springframework/integration/handler/LoggingHandler.java]: That's exactly I've written you about :smile: Let's implement this solution and get rid of `onInit` method at all.

File: spring-integration-core/src/main/java/org/springframework/integration/handler/LoggingHandler.java
Comment (englishman) [inline spring-integration-core/src/main/java/org/springframework/integration/handler/LoggingHandler.java]: Agreed

File: spring-integration-core/src/test/java/org/springframework/integration/handler/LoggingHandlerTests.java
Comment (englishman) [inline spring-integration-core/src/test/java/org/springframework/integration/handler/LoggingHandlerTests.java]: Here is better to use JUnit [ExpectedException rule](http://junit.org/junit4/javadoc/4.12/org/junit/rules/ExpectedException.html)

File: spring-integration-core/src/main/java/org/springframework/integration/handler/LoggingHandler.java
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/handler/LoggingHandler.java]: No, it's not funny.
The `onInit()` must be there anyway. And its purpose to provide `BeanFactoryAware` `expression` and `evaluationContext`. 
I'm not sure which side effects might happen if you use the `LoggingHandler` without `BeanFactory`, but as you insist we should reinstate the backward compatibility for this components. As Gary says: cat is out of bag already. We can only deprecate such a wrong behavior.
But seems for me we have spent already so much time for so simple problem.

If I would do a new component I advocate just only for the  `BeanFactoryAware` behavior, but since we have a case when some code is already used it improperly and there is no so easy way to change that code so, we decided to meet your requirements and reinstate an old behavior.
That's all. I don't see any other arguments to discuss further.

File: spring-integration-core/src/test/java/org/springframework/integration/handler/LoggingHandlerTests.java
Comment (artembilan) [inline spring-integration-core/src/test/java/org/springframework/integration/handler/LoggingHandlerTests.java]: Maybe, but not in all cases...
Especially when I would like to verify several calls in a single test-case.

Comment (englishman) [inline spring-integration-core/src/test/java/org/springframework/integration/handler/LoggingHandlerTests.java]: Provide a new test

File: spring-integration-core/src/main/java/org/springframework/integration/handler/LoggingHandler.java
Comment (englishman) [inline spring-integration-core/src/main/java/org/springframework/integration/handler/LoggingHandler.java]: Agreed. Changed. Pls, review.

File: spring-integration-core/src/test/java/org/springframework/integration/handler/LoggingHandlerTests.java
Comment (artembilan) [inline spring-integration-core/src/test/java/org/springframework/integration/handler/LoggingHandlerTests.java]: Great work!
Thank you!

Minor here:
No, we don't accept asterisk imports: http://stackoverflow.com/questions/147454/why-is-using-a-wild-card-with-a-java-import-statement-bad

Comment (artembilan) [inline spring-integration-core/src/test/java/org/springframework/integration/handler/LoggingHandlerTests.java]: Right, and Checkstyle says you about that:
```
:spring-integration-core:checkstyleTest[ant:checkstyle] [ERROR] /home/travis/build/spring-projects/spring-integration/spring-integration-core/src/test/java/org/springframework/integration/handler/LoggingHandlerTests.java:19: Using the '.*' form of import should be avoided - org.junit.Assert.*. [AvoidStarImport]

[ant:checkstyle] [ERROR] /home/travis/build/spring-projects/spring-integration/spring-integration-core/src/test/java/org/springframework/integration/handler/LoggingHandlerTests.java:20: Using the '.*' form of import should be avoided - org.mockito.Mockito.*. [AvoidStarImport]
```
For this purpose you can run testing like `gradlew clean check --parallel`. 
Sometime if I need really only checkstyle, I add `-x test` flag as well.

I'm waiting for your fix and merge...

Comment (englishman) [inline spring-integration-core/src/test/java/org/springframework/integration/handler/LoggingHandlerTests.java]: @artembilan fixed. BTW I've used standard [IntelliJ IDEA code format settings](https://github.com/spring-projects/spring-framework/tree/master/src/idea). Them should be fixed then. I'll provide needed changes.

**Review by artembilan — APPROVED** [review]

**Comment by artembilan** [comment]
Please, fix your Git client for your real name. `englishman <kontiky@gmail.com>` doesn't look appropriate.
This is Open Source, so everything should be transparent. You have signed CLA therefore the code with your name should go according those rules.

Thank you for understanding.

The current commit I'll fix myself, because I know your name, but for the future it would be more convenient to be automated.

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-integration/commit/ee848e9ecf938010cd8b39d4aa251dced8472735 and cherry-picked.

Thank you for contribution.
Looking forward for more!

**Pull request closed without merging**
